cpp/src/io/parquet/reader_impl.cpp

```diff
@@ -122,7 +159,16 @@ void reader::impl::decode_page_data(size_t skip_rows, size_t num_rows)
   chunk_nested_valids.host_to_device_async(_stream);
   chunk_nested_data.host_to_device_async(_stream);
 
-  gpu::DecodePageData(pages, chunks, num_rows, skip_rows, _file_itm_data.level_type_size, _stream);
+  auto stream1 = _stream_pool.get_stream();
+  gpu::DecodePageData(pages, chunks, num_rows, skip_rows, _file_itm_data.level_type_size, stream1);
+  if (has_strings) {
+    auto stream2 = _stream_pool.get_stream();
+    chunk_nested_str_data.host_to_device_async(stream2);
+    gpu::DecodeStringPageData(
+      pages, chunks, num_rows, skip_rows, _file_itm_data.level_type_size, stream2);
+    stream2.synchronize();
+  }
+  stream1.synchronize();
```

Comment:
I think we should always guarantee that we call the first kernel with the passed in stream. So stream1 here should always just be `_stream`.  Then we can pull off other streams from the pool as needed.  

Also, I think we need a `_stream.synchronize()` before we launch these kernels because of the `host_to_device_async()` calls above.

Reply:
As feared, using `_stream` for the first kernel tanks performance for the small mixed case. Is this ok?
```
## parquet_read_io_small_mixed

pooled stream1

|    io    | cardinality | run_length | num_string_cols | Samples | CPU Time | Noise | GPU Time | Noise | bytes_per_second | peak_memory_usage | encoded_file_size |
|----------|-------------|------------|-----------------|---------|----------|-------|----------|-------|------------------|-------------------|-------------------|
| FILEPATH |        1000 |          1 |               2 |   1312x | 2.267 ms | 7.41% | 2.262 ms | 7.41% |     237354738759 |        10.487 MiB |         1.019 MiB |
| FILEPATH |        1000 |         32 |               2 |   2544x | 1.604 ms | 2.02% | 1.599 ms | 1.97% |     335678427608 |         9.585 MiB |       126.455 KiB |

_stream for first kernel
|    io    | cardinality | run_length | num_string_cols | Samples | CPU Time | Noise | GPU Time | Noise | bytes_per_second | peak_memory_usage | encoded_file_size |
|----------|-------------|------------|-----------------|---------|----------|-------|----------|-------|------------------|-------------------|-------------------|
| FILEPATH |        1000 |          1 |               2 |   2064x | 3.361 ms | 2.30% | 3.356 ms | 2.29% |     159987070458 |        10.487 MiB |         1.019 MiB |
| FILEPATH |        1000 |         32 |               2 |   1280x | 1.920 ms | 2.04% | 1.915 ms | 1.98% |     280400202774 |         9.585 MiB |       126.455 KiB |
```

Reply:
@nvdbaranec why do you think that the first kernel should use the user stream? IMO it's fine to fork to different streams in the middle of the read call.
Although, the more I think about this, the less sense I can make of the user stream parameter overall (given that read_parquet is synchronous in practice).